### PR TITLE
refactor to introduce local struct

### PR DESCRIPTION
**Pull request by jcantrill**
This change is to propose address https://jira.coreos.com/browse/LOG-357:

* introduction of a local struct to allow use of attaching methods
* Add code style based on golang conventions
* Improve readability and promote code reuse

### CODE REVIEW

File: pkg/k8shandler/collection.go [hunk 1]
```diff
@@ -5,36 +5,42 @@ import (
 	"io/ioutil"
 	"reflect"
 
+	"strings"
+	"time"
+
 	"github.com/openshift/cluster-logging-operator/pkg/utils"
 	"github.com/sirupsen/logrus"
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (ewolinetz) [inline pkg/k8shandler/collection.go]: fyi this is unnecessary, the defaulted name it gets is the last part of the package which is already "v1"

File: pkg/k8shandler/collection.go [hunk 2]
```diff
@@ -5,36 +5,42 @@ import (
...
 var (
 	retryInterval = time.Second * 30
 	timeout       = time.Second * 1800
 )
 
-func CreateOrUpdateCollection(cluster *logging.ClusterLogging) (err error) {
+func (localLogging *ClusterLogging) CreateOrUpdateCollection() (err error) {
 
-	if err = createOrUpdateCollectionPriorityClass(cluster); err != nil {
+	cluster := localLogging.ClusterLogging
```
Comment (ewolinetz) [inline pkg/k8shandler/collection.go]: if you update the function to be for `(cluster *logging.ClusterLogging)` does this line become unnecessary?

File: pkg/k8shandler/types.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+}
+
+//DoesClusterLoggingExist determines if the specified cluster exists
+func (clusterLogging *ClusterLogging) DoesClusterLoggingExist() bool {
+	exists, _ := utils.DoesClusterLoggingExist(clusterLogging.ClusterLogging)
+	return exists
+}
+
+//createOrUpdateServiceAccount creates or updates a ServiceAccount for logging with the given name
+func (logging *ClusterLogging) createOrUpdateServiceAccount(name string) error {
```
Comment (ewolinetz) [inline pkg/k8shandler/types.go]: I would argue this belongs in `utils` and the functions that return an object definition would belong in here.

File: pkg/k8shandler/collection.go (hunk 1 above)
Comment (jcantrill) [reply pkg/k8shandler/collection.go]: ack.  this is copy/past from some other file

File: pkg/k8shandler/collection.go (hunk 2 above)
Comment (jcantrill) [reply pkg/k8shandler/collection.go]: This is really a 'stop gap' to make refactoring more digestible.  I would expect this line to be unnecessary after refactoring all the places where its needed

File: pkg/k8shandler/collection.go (hunk 1 above)
Comment (jcantrill) [reply pkg/k8shandler/collection.go]: interestingly, this is being reverted for me.  I wonder if there is some intelligence to 'go fmt' that is adding it back in

**Comment by jcantrill** [comment]
@jcantrill   Here is my 2 cents about the PR as implemented by adding methods to the `pkg/api` folder:
* It basically requires you to more or less move all `cluster.foo()` logic which was in k8shandler to the `pkg/api` folder.  That seems less then desirable but probably a symmantic/style difference.  
* The meat of the logic which is in k8shandler also now becomes versioned because `pkg/api` is versioned.  This may make future updates non-intuative and or more difficult.  One solution would be to introduce interfaces but I don't see how we would need multiple versions from version to version given the logic probably wont drift much
* A better approach might be to introduce a simple wrapper object simply for the reason you can create it with the internal as an anonymous type and all fields just 'bubble' up.  It may also make it cleaner for us to migrate versions in future.

**Comment by jcantrill** [comment]
@richm @ewolinetz   I have added a second commit based on my previous comment and recent experience with the code base.  It's my opinion we should work towards the newest comment in lieu of the first commit; it uses a 'wrapper' struct and allows us to keep the pks/api package clean.  This seems to fit into the pattern of other projects and operators as best as I can determine.  I personally don't find it any more challenging and think it will allow us to better consume version changes as well as evolve the 'private' api without needing to modify the 'public' api

**Comment by richm** [comment]
lgtm but I don't really understand the motivation of this refactoring

**Comment by jcantrill** [comment]
> lgtm but I don't really understand the motivation of this refactoring

The main motivations are:
* Improve readability
* Encapsulate certain logic, OO style in places where it makes sense in lieu of relying on functional aspects
* Collapse duplicate logic into reusable functions

My belief is by doing so the code will be cleaner, easier to understand, and maintain

**Comment by ewolinetz** [comment]
/hold
I would argue that since this doesn't fix anything directly we should
refrain from merging it yet during a freeze.

File: pkg/stub/handler.go [hunk 4]
```diff
@@ -34,49 +33,49 @@ func (h *Handler) Handle(ctx context.Context, event sdk.Event) error {
...
 }
 
-func Reconcile(cluster *logging.ClusterLogging) (err error) {
+func Reconcile(spec *logging.ClusterLogging) (err error) {
 	exists := true
-
+	cluster := k8shandler.NewClusterLogging(spec)
 	// Reconcile certs
-	if exists, cluster = utils.DoesClusterLoggingExist(cluster); exists {
+	if exists = cluster.Exists(); exists {
```
Comment (ewolinetz) [inline pkg/stub/handler.go]: we can just shorten this to `if cluster.Exists() {`

Comment (ewolinetz) [reply pkg/stub/handler.go]: That previous function also served to get us the most recent iteration of the cluster object so that we could update it without conflicts... does this still cover that case?

Comment (jcantrill) [reply pkg/stub/handler.go]: `cluster` is  a pointer meaning the data associated with the struct to which it points is mutable.  I followed the `sdk.Get` function and the way I interpreted it was that it would 'fill' the struct identified by the pointer. If read correctly then yes its covered.  I would prefer an additional set of eyes to ensure it behaves as I expect.

File: pkg/utils/factory.go [hunk 5]
```diff
@@ -44,6 +47,9 @@ func Secret(secretName string, namespace string, data map[string][]byte) *core.S
 	}
 }
 
+func NewServiceAccount(accountName string, namespace string) *core.ServiceAccount {
+	return ServiceAccount(accountName, namespace)
```
[the remark this answers is not in the document]
Comment (jcantrill) [reply pkg/utils/factory.go]: It is only a stop gap until we move ALL the code to use `New....`.   The golang style convention is prefix functions with `New` if they represent constructors, which this is.  We should move to golang conventions as much as possible.

File: pkg/utils/utils.go [hunk 6]
```diff
@@ -39,12 +32,26 @@ var COMPONENT_IMAGES = map[string]string{
...
+	Meta() metav1.ObjectMeta
+}
+
+func AllInOne(o KubernetesResource) bool {
+
+	//_, ok := logging.ObjectMeta.Annotations[SPLIT_ANNOTATION]
+	return true
+}
+
+func AsOwner(o KubernetesResource) metav1.OwnerReference {
```
[the remark this answers is not in the document]
Comment (jcantrill) [reply pkg/utils/utils.go]: I think this is a standin until KubernetesResource goes away to minimize impact of the other places change would need to be made

**Comment by jcantrill** [comment]
@ewolinetz rebased to include latest and refactored 'Exists()' to check based on a copy

**Comment by jcantrill** [comment]
block PR on https://github.com/openshift/cluster-logging-operator/pull/105

**Comment by richm** [comment]
registry issue fixed - now running into the usual "no more aws resources - cannot install" issue
/test e2e-aws

File: pkg/apis/logging/v1alpha1/temp.go [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+func DoesClusterLoggingExist(cluster *ClusterLogging) bool {
+
+	// this is to avoid an invalid memory address panic
+	if cluster == nil {
+		return false
+	}
+
+	if err := sdk.Get(cluster); err != nil {
```
Comment (ewolinetz) [inline pkg/apis/logging/v1alpha1/temp.go]: didn't we want to avoid updating the ClusterLogging object? this will do that since we pass it as a pointer.
I think we want to do a deepcopy of it and then `get` on that object

File: pkg/k8shandler/curation.go [hunk 8]
```diff
@@ -48,7 +48,7 @@ func CreateOrUpdateCuration(cluster *logging.ClusterLogging) (err error) {
 
 		printUpdateMessage := true
 		retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
-			if exists, cluster := utils.DoesClusterLoggingExist(cluster); exists {
+			if exists := logging.DoesClusterLoggingExist(cluster); exists {
```
Comment (ewolinetz) [inline pkg/k8shandler/curation.go]: shouldn't this be `cluster.Exists()` ?

File: pkg/apis/logging/v1alpha1/temp.go (hunk 7 above)
Comment (jcantrill) [reply pkg/apis/logging/v1alpha1/temp.go]: We do that here: https://github.com/openshift/cluster-logging-operator/pull/85/files#diff-156742c24dc12665a82a7fed5cf7c8abR57   I'll update in this method too

File: pkg/k8shandler/logstore.go [hunk 9]
```diff
@@ -39,7 +39,7 @@ func CreateOrUpdateLogStore(cluster *logging.ClusterLogging) (err error) {
 
 		printUpdateMessage := true
 		retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
-			if exists, cluster := utils.DoesClusterLoggingExist(cluster); exists {
+			if exists := logging.DoesClusterLoggingExist(cluster); exists {
```
Comment (ewolinetz) [inline pkg/k8shandler/logstore.go]: `cluster.Exists()`?

File: pkg/k8shandler/visualization.go [hunk 10]
```diff
@@ -59,7 +59,7 @@ func CreateOrUpdateVisualization(cluster *logging.ClusterLogging) (err error) {
 
 		printUpdateMessage := true
 		retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
-			if exists, cluster := utils.DoesClusterLoggingExist(cluster); exists {
+			if exists := logging.DoesClusterLoggingExist(cluster); exists {
```
Comment (ewolinetz) [inline pkg/k8shandler/visualization.go]: `cluster.Exists()` ?

File: pkg/k8shandler/visualization.go [hunk 11]
```diff
@@ -83,39 +83,40 @@ func CreateOrUpdateVisualization(cluster *logging.ClusterLogging) (err error) {
 
 func removeKibana(cluster *logging.ClusterLogging) (err error) {
 	if cluster.Spec.ManagementState == logging.ManagementStateManaged {
-		if err = utils.RemoveServiceAccount(cluster, "kibana"); err != nil {
+		namespace := cluster.Namespace
```
Comment (ewolinetz) [inline pkg/k8shandler/visualization.go]: if we do this in the visualization file, can we do this in the other ones too so they look consistent?

File: pkg/k8shandler/curation.go (hunk 8 above)
Comment (jcantrill) [reply pkg/k8shandler/curation.go]: This PR only does a partial refactoring to minimize impact across the code base.  it handles changes related to the collector

File: pkg/k8shandler/visualization.go (hunk 11 above)
Comment (jcantrill) [reply pkg/k8shandler/visualization.go]: yes

**Comment by jcantrill** [comment]
test e2e-aws

File: pkg/k8shandler/collection.go [hunk 12]
```diff
@@ -350,7 +355,7 @@ func createOrUpdateRsyslogConfigMap(logging *logging.ClusterLogging) error {
 	)
 	rsyslogConfigMaps["rsyslog"] = rsyslogConfigMap
 	for name, cm := range rsyslogConfigMaps {
-		utils.AddOwnerRefToObject(cm, utils.AsOwner(logging))
+		logging.AddOwnerRefTo(cm)
```
Comment (ewolinetz) [inline pkg/k8shandler/collection.go]: +1

**Comment by ewolinetz** [comment]
/lgtm
I would like to also eventually condense all api objects to their own files e.g. anything `serviceaccount` related would be in a `serviceaccount.go` etc. That can be a future pr.

Also thanks for removing all the `AllInOne` logic that wasn't being used anymore.
